sogs/model/room.py

```diff
@@ -1243,6 +1243,49 @@ def delete_all_posts(self, poster: User, *, deleter: User):
 
         return len(deleted), files_removed
 
+    def delete_all_user_reactions(self, poster: User, *, deleter: User):
+        """
+        Delete all emoji reactions placed on messages in the room by `poster`.
+        `deleter` must be a moderator if not deleting his/her own reactions, and
+        must be an `admin` if trying to delete all of the posts of another admin.
+        """
+
+        fail = None
+        if poster.id != deleter.id and not self.check_moderator(deleter):
+            fail = "user is not a moderator"
+        elif self.check_admin(poster) and not self.check_admin(deleter):
+            fail = "only admins can delete all reactions of another admin"
+
+        if fail is not None:
+            app.logger.warning(
+                f"Error deleting all reactions by {poster} from {self} by {deleter}: {fail}"
+            )
+            raise BadPermission()
+
+        with db.transaction():
+            deleted = [
+                r[0]
+                for r in query(
+                    'SELECT reaction FROM user_reactions WHERE "user" = :u AND reaction IN (SELECT id FROM reactions WHERE message IN (SELECT id FROM messages WHERE room = :r))',
+                    r=self.id,
+                    u=poster.id,
+                )
+            ]
+
+            query(
+                'DELETE FROM user_reactions WHERE "user" = :u AND reaction IN (SELECT id FROM reactions WHERE message IN (SELECT id FROM messages WHERE room = :r))',
+                r=self.id,
+                u=poster.id,
+            )
```

Comment:
These sorts of queries are messy and are better abstracted away in the SQL definition.

I think these would do it:
```SQL
-- sqlite:
CREATE TRIGGER message_reactions_delete INSTEAD OF DELETE ON message_reactions
FOR EACH ROW
BEGIN
    DELETE FROM user_reactions WHERE reaction = OLD.id AND "user" = OLD."user";
END;

-- postgresql:
CREATE OR REPLACE FUNCTION trigger_message_reactions_delete()
RETURNS TRIGGER LANGUAGE PLPGSQL AS $$BEGIN
    DELETE FROM user_reactions WHERE reaction = OLD.id AND "user" = OLD."user";
    RETURN NULL;
END;$$;
CREATE TRIGGER message_reactions_delete INSTEAD OF DELETE ON message_reactions
FOR EACH ROW
EXECUTE PROCEDURE trigger_message_reactions_delete();
```

And then the query executed in the code here can just be:
```SQL
DELETE FROM message_reactions WHERE "user" = :u
```

(Those triggers need to go into schema.sqlite/scheme.pgsql for new installs, and into a migration script for upgrades--I'll look into how to properly query checking for that as it's not as simple as most of the existing code that just looks for the existence of a table/column to figure out when it needs to migrate.)

Reply:
Wait, no, that's not right.

Reply:
I think maybe we should create another view for these cases where we need both reaction + room.

```sql
-- This should work for both sqlite/pgsql:
CREATE VIEW room_reactions AS
SELECT reactions.*, user_reactions."user", user_reactions.at, messages.room
FROM user_reactions
    JOIN reactions ON user_reactions.reaction = reactions.id
    JOIN messages ON reactions.message = messages.id;

-- sqlite:
CREATE TRIGGER room_reactions_delete INSTEAD OF DELETE ON room_reactions
FOR EACH ROW
BEGIN
    DELETE FROM user_reactions WHERE reaction = OLD.id AND "user" = OLD."user";
END;

-- postgresql:
CREATE OR REPLACE FUNCTION trigger_room_reactions_delete()
RETURNS TRIGGER LANGUAGE PLPGSQL AS $$BEGIN
    DELETE FROM user_reactions WHERE reaction = OLD.id AND "user" = OLD."user";
    RETURN NULL;
END;$$;
CREATE TRIGGER room_reactions_delete INSTEAD OF DELETE ON room_reactions
FOR EACH ROW
EXECUTE PROCEDURE trigger_room_reactions_delete();
```

Then queries that want to be room specific can use this view.  So for instance:

```sql
DELETE FROM room_reactions WHERE room = :r AND "user" = :u;
```
should accomplish what we want here.

Reply:
As for where to put this:

- The definitions go into schema.sqlite and schema.pgsql, for new installs.
- Upgrade code to add the new definitions needs to go into a new migration code in sogs/migration (there's a bunch of code to use an examples there now; `reactions.py` is probably a simple enough one to copy into a new one.  Basically:
    - Add a new file with a `def migrate(...)` that does the check-or-throw for whether the thing being added already exists and whether we're in check-only mode.
    - For sqlite, the migration additions have to be done as separate `conn.execute` statements
    - For postgresql it is nicer to use a single multi-query statement to add them exactly as is done in the schema.pgsql
    - the new reaction file gets added into `migrations/__init__.py` to actually get called on upgrade

Reply:
(There are probably a couple other cases in the codebase where using this view would simplify queries, if you're feeling bold, but don't consider it necessary for this PR).

Reply:
(Also if you're wondering why not just add the room into the `message_reactions` view: the extra JOIN isn't free and can't be optimized away from the query by the SQL engine just because you don't use the value because a JOIN potentially changes the set of rows that get returned).